sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java

```diff
@@ -321,6 +322,117 @@ public void queryMetrics(String query, Boolean qmEnabled) {
+        CosmosClient testDirectClient = new CosmosClientBuilder()
+                                            .endpoint(TestConfigurations.HOST)
+                                            .key(TestConfigurations.MASTER_KEY)
+                                            .contentResponseOnWriteEnabled(true)
+                                            .directMode()
+                                            .buildClient();
+        CosmosQueryRequestOptions options = new CosmosQueryRequestOptions();
+
+        CosmosContainer cosmosContainer = testDirectClient.getDatabase(cosmosAsyncContainer.getDatabase().getId())
+                                              .getContainer(cosmosAsyncContainer.getId());
+        List<String> itemIdList = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            InternalObjectNode internalObjectNode = getInternalObjectNode();
+            CosmosItemResponse<InternalObjectNode> createResponse = cosmosContainer.createItem(internalObjectNode);
+            if (i % 20 == 0) {
+                itemIdList.add(internalObjectNode.getId());
+            }
+        }
+        boolean qroupByFirstResponse = true;
+        if (qmEnabled != null) {
+            options.setQueryMetricsEnabled(qmEnabled);
+        }
+        Iterator<FeedResponse<InternalObjectNode>> iterator = cosmosContainer
+                                                                  .queryItems(query, options, InternalObjectNode.class)
+                                                                  .iterableByPage()
+                                                                  .iterator();
+        assertThat(iterator.hasNext()).isTrue();
+
+        while (iterator.hasNext()) {
+            FeedResponse<InternalObjectNode> feedResponse = iterator.next();
+            String queryDiagnostics = feedResponse.getCosmosDiagnostics().toString();
+            assertThat(feedResponse.getResults().size()).isEqualTo(0);
+            if (!query.contains("group by") || qroupByFirstResponse) { // TODO https://github
+                if (qmEnabled) {
+                    validateQueryDiagnostics(queryDiagnostics, qmEnabled, true);
+                    validateDirectModeQueryDiagnostics(queryDiagnostics);
+                    validateTransportRequestTimelineDirect(queryDiagnostics);
+                }
+                if (query.contains("group by")) {
+                    qroupByFirstResponse = false;
+                }
+            }
+        }
+    }
+
+    private void validateDirectModeQueryDiagnostics(String diagnostics) {
+        assertThat(diagnostics).contains("\"connectionMode\":\"DIRECT\"");
+        assertThat(diagnostics).contains("supplementalResponseStatisticsList");
+        assertThat(diagnostics).contains("\"gatewayStatistics\":null");
+        assertThat(diagnostics).contains("addressResolutionStatistics");
+        assertThat(diagnostics).contains("\"userAgent\":\"" + Utils.getUserAgent() + "\"");
```

Comment:
This is very limited for validation.

this doesn't really structurally validates what is in the diagnostics. please take a look at `CilentConfigDiagnosticsTest` in some cases it is worth fully validate what goes in the diagnostics. the number of entries, order, etc.

Reply:
Individual request diagnostics are already being validated as part of store diagnostic tests. Additional validation on query with counts etc is bit more involved as that depends on current consistency, partition count etc. I will try to add those eventually